projects/igniteui-angular/src/lib/services/transaction/igx-transaction.ts

```diff
@@ -129,35 +135,24 @@ export class IgxTransactionService<T extends Transaction, S extends State> exten
             return;
         }
 
-        let action: { transaction: T, recordRef: any, useInUndo?: boolean };
-        do {
-            action = this._undoStack.pop();
-            this._transactions.pop();
-            this._redoStack.push(action);
-        } while (!action.useInUndo);
+        const actions: { transaction: T, recordRef: any }[] = this._undoStack.pop();
+        this._transactions.splice(this._transactions.length - actions.length);
+        this._redoStack.push(actions);
 
         this._states.clear();
-        this._undoStack.map(a => this.updateState(this._states, a.transaction, a.recordRef));
+        this._undoStack.map(a => a.map(t => this.updateState(this._states, t.transaction, t.recordRef)));
         this.onStateUpdate.emit();
     }
 
     public redo(): void {
         if (this._redoStack.length > 0) {
-            //  remove first item from redo stack (it should always has useInUndo === true)
-            //  and then all next items until there are items and useInUndo === false.
-            //  If there are no more items, or next item's useInUndo === true leave.
-            let undoItem: { transaction: T, recordRef: any, useInUndo?: boolean };
-            undoItem = this._redoStack.pop();
-            this.updateState(this._states, undoItem.transaction, undoItem.recordRef);
-            this._transactions.push(undoItem.transaction);
-            this._undoStack.push(undoItem);
-
-            while (this._redoStack[this._redoStack.length - 1] && !this._redoStack[this._redoStack.length - 1].useInUndo) {
-                undoItem = this._redoStack.pop();
-                this.updateState(this._states, undoItem.transaction, undoItem.recordRef);
-                this._transactions.push(undoItem.transaction);
-                this._undoStack.push(undoItem);
-            }
+            let actions: { transaction: T, recordRef: any, useInUndo?: boolean }[];
+            actions = this._redoStack.pop();
+            actions.map(a => {
```

Comment:
That's a map used as `forEach`. For-of will do too.

Reply:
yes, same in redo function - fixed now